File: ios/PSMobile/Screens/SetupView.swift
import SwiftUI
import PSMShared

/// Ersteinrichtung - Gegenstueck zum Konfigurationsassistenten des
/// Desktops und zu SetupScreen.kt auf Android.
///
/// Der Zweck ist nicht Kosmetik: Werden alle 37 Druckermodelle
/// installiert, dauert der Start 13,5 s und die Auswahllisten enthalten
/// 5762 Filamente. Mit einem einzelnen gewaehlten Drucker sind es 1,9 s
/// und 189 Filamente.
///
/// Die Gruppierung nach Familien kommt aus dem gemeinsamen Modul
/// (E-13) - dieselbe Reihenfolge wie auf Android, dieselbe Regel fuer
/// Altgeraete am Ende.
struct SetupView: View {

    let models: [PsmCore.PrinterModel]
    let busy: Bool
    /// Bisher installierte Modelle, Format "vendor:model:variant".
    let preselected: Set<String>
    let onConfirm: ([String]) -> Void
    let onLanguageChange: (String) -> Void
    var onClose: (() -> Void)? = nil

    @Environment(\.psScale) private var ps
    @State private var selected: Set<String> = []
    @State private var query = ""
    @State private var expandedVendors: Set<String> = []
    @State private var expandedFamilies: Set<String> = []

    /// Ist es eng? Dann faellt Beiwerk weg, statt alles zu schrumpfen.
    /// Dieselbe Frage wie UiScale.density auf Android; die Schwelle
    /// stammt von dort.
    private var tight: Bool { ps.factor <= 0.8 }

    private var shown: [PsmCore.PrinterModel] {
        models.filter { m in
            // Hart aus, nicht nur standardmaessig versteckt: PSMobile
            // schneidet nur FFF (E-08), und das mitgelieferte
            // PrusaResearchSLA-Buendel ist zudem kaputt (siehe
            // build/scripts/stage-resources.sh). Ein Schalter, der zu
            // einem funktionsunfaehigen Drucker fuehrt, ist schlimmer
            // als gar keiner.
            !m.isSla &&
            (query.trimmingCharacters(in: .whitespaces).isEmpty ||
             m.name.localizedCaseInsensitiveContains(query))
        }
    }

    /// Eine Familie mit ihren Modellen.
    ///
    /// Die Regel im gemeinsamen Modul liefert Positionen statt Modelle -
    /// so ueberlebt sie die Bruecke, denn Kotlins Typparameter gehen nach
    /// Swift verloren. Hier zurueck zu den eigenen Objekten.
    private struct Familie: Identifiable {
        let name: String
        let isLegacy: Bool
        let modelle: [PsmCore.PrinterModel]
        var id: String { name }
    }

    /// Ein Hersteller mit seinen Familien.
    private struct Hersteller: Identifiable {
        let name: String
        let familien: [Familie]
        var anzahl: Int { familien.reduce(0) { $0 + $1.modelle.count } }
        var id: String { name }
    }

    /// Zwei Ebenen: Hersteller, darunter Familien.
    ///
    /// Beide Einteilungen kommen aus dem gemeinsamen Modul und beide
    /// aus PrusaSlicers eigenen Daten - der Hersteller aus dem
    /// Schluessel, die Familie aus der Vendor-Datei.
    private var hersteller: [Hersteller] {
        let liste = shown
        return PrinterGrouping.shared.groupByVendor(keys: liste.map { $0.key })
            .map { v in
                let modelle = v.indices.map { liste[$0.intValue] }
                let familien = PrinterGrouping.shared
                    .group(families: modelle.map { $0.family })
                    .map { g in
                        Familie(name: g.family,
                                isLegacy: g.isLegacy,
                                modelle: g.indices.map { modelle[$0.intValue] })
                    }
                return Hersteller(name: v.family, familien: familien)
            }
    }

    var body: some View {
        SchwebenderDialog(kennung: "dialog.einrichtung", maximaleBreite: ps.pt(760)) {
            VStack(alignment: .leading, spacing: ps.pt(tight ? 6 : 14)) {
                kopfzeile
                suchfeld
                liste
                abschluss
            }
            .padding(ps.pt(tight ? 12 : 24))
            .frame(maxHeight: .infinity)
        }
        .onAppear { selected = preselected }
    }

    // MARK: - Teile

    private var kopfzeile: some View {
        HStack(alignment: .center) {
            if let onClose {
                Button(action: onClose) {
                    Text("‹  " + SimpleModeState.shared.text(english: "Back", german: "Zurück"))
                        .font(.system(size: ps.font(15)))
                        .foregroundStyle(PrusaColors.orange)
                        .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .accessibilityIdentifier("einrichtung.schliessen")
            }
            VStack(alignment: .leading, spacing: ps.pt(2)) {
                Text(SimpleModeState.shared.text(english: "Configuration Assistant",
                                                 german: "Ersteinrichtung"))
                    .font(.system(size: ps.font(tight ? 18 : 22), weight: .semibold))
                    .foregroundStyle(PrusaColors.textPrimary)

                // Auf engen Schirmen faellt der Untertitel weg. Er
                // erklaert nichts, was die Liste darunter nicht selbst
                // zeigt, kostet aber eine Zeile, die dann der Liste
                // fehlt.
                if !tight {
                    Text(SimpleModeState.shared.text(
                            english: "Select all printers you want to use.",
                            german: "Waehle alle Drucker, die du nutzen willst."))
                        .font(.system(size: ps.font(13)))
                        .foregroundStyle(PrusaColors.textMuted)
                }
            }
            Spacer()
        }
    }

    private var suchfeld: some View {
        // Der Platzhalter wird selbst gezeichnet. SwiftUI faerbt seinen
        // eigenen in einem Grau, das auf dem dunklen Feld praktisch
        // unsichtbar ist - das Feld sah aus wie ein leerer Kasten ohne
        // Hinweis, wozu er da ist.
        ZStack(alignment: .leading) {
            if query.isEmpty {
                Text(SimpleModeState.shared.text(english: "Search printer model",
                                                 german: "Druckermodell suchen"))
                    .font(.system(size: ps.font(15)))
                    .foregroundStyle(PrusaColors.textMuted)
                    .padding(.horizontal, ps.pt(12))
                    .allowsHitTesting(false)
            }
            TextField("", text: $query)
                .textFieldStyle(.plain)
                .font(.system(size: ps.font(15)))
                .foregroundStyle(PrusaColors.textPrimary)
                .padding(ps.pt(12))
                .autocorrectionDisabled()
        }
        .background(PrusaColors.panelRaised)
        .clipShape(RoundedCornerShape(ps.pt(6)))
    }

    private var liste: some View {
        let alle = hersteller
        return ScrollView {
            LazyVStack(alignment: .leading, spacing: 0) {
                ForEach(alle) { marke in
                    // Bei nur einem Hersteller waere die oberste Ebene
                    // eine Zeile, die man immer erst aufklappen muss,
                    // ohne dass sie etwas unterscheidet.
                    let offen = alle.count == 1
                        || expandedVendors.contains(marke.name)
                        || !query.isEmpty
                    if alle.count > 1 {
                        herstellerKopf(marke, offen: offen)
                    }
                    if offen {
                        ForEach(marke.familien) { gruppe in
                            familienKopf(gruppe)
                            if expandedFamilies.contains(gruppe.name) || !query.isEmpty {
                                ForEach(gruppe.modelle, id: \.key) { modell in
                                    modellZeile(modell)
                                }
                            }
                        }
                    }
                }
            }
        }
        .frame(maxHeight: .infinity)
    }

    private func herstellerKopf(_ marke: Hersteller, offen: Bool) -> some View {
        Button {
            if expandedVendors.contains(marke.name) {
                expandedVendors.remove(marke.name)
            } else {
                expandedVendors.insert(marke.name)
            }
        } label: {
            HStack {
                Text(marke.name)
                    .font(.system(size: ps.font(17), weight: .bold))
                    .foregroundStyle(PrusaColors.textPrimary)
                Text("\(marke.anzahl)")
                    .font(.system(size: ps.font(13)))
                    .foregroundStyle(PrusaColors.textMuted)
                Spacer()
                Image(systemName: offen ? "chevron.down" : "chevron.right")
                    .font(.system(size: ps.font(14)))
                    .foregroundStyle(PrusaColors.textMuted)
            }
            .padding(.top, ps.pt(8))
            .frame(minHeight: ps.touch(52))
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("hersteller.\(marke.name)")
    }

    private func familienKopf(_ gruppe: Familie) -> some View {
        Button {
            if expandedFamilies.contains(gruppe.name) {
                expandedFamilies.remove(gruppe.name)
            } else {
                expandedFamilies.insert(gruppe.name)
            }
        } label: {
            HStack {
                Text(gruppe.name.uppercased())
                    .font(.system(size: ps.font(14), weight: .semibold))
                    .padding(.leading, ps.pt(12))
                    .foregroundStyle(gruppe.isLegacy
                                     ? PrusaColors.textMuted : PrusaColors.orange)
                Text("\(gruppe.modelle.count)")
                    .font(.system(size: ps.font(13)))
                    .foregroundStyle(PrusaColors.textMuted)
                Spacer()
                Image(systemName: expandedFamilies.contains(gruppe.name)
                      ? "chevron.down" : "chevron.right")
                    .font(.system(size: ps.font(14)))
                    .foregroundStyle(PrusaColors.textMuted)
            }
            .frame(minHeight: ps.touch(52))
            // Ohne das ist nur der Text antippbar, nicht die Zeile: der
            // Spacer dazwischen ist leerer Raum, und leeren Raum nimmt
            // SwiftUI von der Trefferpruefung aus. Wer auf die Mitte der
            // Zeile tippt, greift ins Leere - der UI-Test hat es gefunden,
            // ein Finger haette dasselbe erlebt.
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        // Kennungen, damit der UI-Test die Zeilen findet. Der sichtbare
        // Text taugt dafuer nicht: er ist uebersetzt und geht bei der
        // naechsten Sprachaenderung kaputt.
        .accessibilityIdentifier("familie.\(gruppe.name)")
    }

    private func modellZeile(_ modell: PsmCore.PrinterModel) -> some View {
        // Jede Duesengroesse ist eine eigene Wahl - PrusaSlicer fuehrt sie
        // als getrennte Profile, und wer eine 0.6er Duese hat, braucht
        // nicht die 0.4er Profile.
        HStack(alignment: .top, spacing: ps.pt(10)) {
            druckerBild(modell)

            VStack(alignment: .leading, spacing: ps.pt(4)) {
                Text(modell.name)
                    .font(.system(size: ps.font(tight ? 16 : 17)))
                    .foregroundStyle(PrusaColors.textPrimary)

                HStack(spacing: ps.pt(6)) {
                    ForEach(modell.variants, id: \.self) { variante in
                        let key = "\(modell.key):\(variante)"
                        Button {
                            if selected.contains(key) { selected.remove(key) }
                            else { selected.insert(key) }
                        } label: {
                            Text(variante)
                                .font(.system(size: ps.font(15)))
                                .foregroundStyle(selected.contains(key)
                                                 ? .white : PrusaColors.textMuted)
                                .padding(.horizontal, ps.pt(16))
                                .frame(minWidth: ps.touch(64),
                                       minHeight: ps.touch(tight ? 48 : 56))
                                .background(selected.contains(key)
                                            ? PrusaColors.orange : PrusaColors.panelRaised)
                                .clipShape(RoundedCornerShape(ps.pt(4)))
                        }
                        .buttonStyle(.plain)
                        .accessibilityIdentifier("variante.\(key)")
                    }
                }
            }
        }
        .padding(.vertical, ps.pt(tight ? 10 : 14))
    }

    /// Vorschaubild je Druckermodell, wie in der Desktop-App - liegt als
    /// loses PNG in den gebuendelten Ressourcen (profiles/<vendor>/
    /// <model_id>_thumbnail.png), nicht im Asset-Katalog, deshalb ueber
    /// den Dateipfad statt per Image(_:)-Name geladen. Fehlt die Datei
    /// (z.B. bei einem Vendor ohne Vorschaubilder), bleibt ein
    /// Platzhalter-Symbol statt einer leeren Luecke.
    private func druckerBild(_ modell: PsmCore.PrinterModel) -> some View {
        let seite = ps.pt(tight ? 40 : 48)
        return Group {
            if let pfad = thumbnailPfad(modell), let bild = UIImage(contentsOfFile: pfad) {
                Image(uiImage: bild)
                    .resizable()
                    .aspectRatio(contentMode: .fit)
            } else {
                Image(systemName: "printer")
                    .font(.system(size: ps.font(18)))
                    .foregroundStyle(PrusaColors.textMuted)
            }
        }
        .frame(width: seite, height: seite)
        .background(PrusaColors.panelRaised)
        .clipShape(RoundedCornerShape(ps.pt(4)))
    }

    private func thumbnailPfad(_ modell: PsmCore.PrinterModel) -> String? {
        // modell.key hat die Form "vendor:model_id" - der Dateiname der
        // Vorschau folgt bei den meisten Herstellern genau der model_id
        // (dem [printer_model:XXX]-Abschnitt in der Vendor-INI). Voron
        // ist die Ausnahme: dort heisst die Datei z.B.
        // "Voron_v2_250_thumbnail.png" statt "V2_250_thumbnail.png" -
        // Vendor-Praefix und andere Gross-/Kleinschreibung. Deshalb erst
        // den genauen Namen versuchen, sonst im Vendor-Ordner nach einer
        // "*_thumbnail.png" suchen, die die model_id enthaelt.
        let teile = modell.key.split(separator: ":", maxSplits: 1)
        guard teile.count == 2 else { return nil }
        guard let resDir = Bundle.main.resourceURL?.appendingPathComponent("psresources") else {
            return nil
        }
        let vendorDir = resDir
            .appendingPathComponent("profiles")
            .appendingPathComponent(String(teile[0]))
        let genau = vendorDir.appendingPathComponent("\(teile[1])_thumbnail.png")
        if FileManager.default.fileExists(atPath: genau.path) {
            return genau.path
        }
        let modelId = String(teile[1]).lowercased()
        guard let dateien = try? FileManager.default.contentsOfDirectory(
            at: vendorDir, includingPropertiesForKeys: nil) else { return nil }
        return dateien.first {
            let n = $0.lastPathComponent.lowercased()
            return n.hasSuffix("_thumbnail.png") && n.contains(modelId)
        }?.path
    }

    private var abschluss: some View {
        Button {
            onConfirm(Array(selected))
        } label: {
            HStack {
                Spacer()
                if busy {
                    ProgressView().tint(.white)
                } else {
                    Text(SimpleModeState.shared.text(english: "Finish",
                                                     german: "Fertig"))
                        .font(.system(size: ps.font(16), weight: .semibold))
                }
                Spacer()
            }
            .frame(height: ps.touch(tight ? 46 : 58))
            .background(selected.isEmpty ? PrusaColors.panelRaised : PrusaColors.orange)
            .foregroundStyle(selected.isEmpty ? PrusaColors.textMuted : .white)
            .clipShape(RoundedCornerShape(ps.pt(6)))
        }
        .buttonStyle(.plain)
        .disabled(selected.isEmpty || busy)
        .accessibilityIdentifier("fertig")
    }
}

/// Abkuerzung, weil RoundedRectangle(cornerRadius:) sich hier zwanzigmal
/// wiederholt haette.
private func RoundedCornerShape(_ radius: CGFloat) -> RoundedRectangle {
    RoundedRectangle(cornerRadius: radius)
}

File: android/app/src/main/java/de/psmobile/ui/SetupView.kt
package de.psmobile.ui

import android.content.Context
import android.graphics.BitmapFactory
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import de.psmobile.core.PsmCore
import de.psmobile.shared.rules.PrinterGrouping
import de.psmobile.slicing.ResourceInstaller
import de.psmobile.ui.theme.PrusaColors
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

/// Ersteinrichtung - Gegenstueck zum Konfigurationsassistenten des
/// Desktops.
///
/// Der Zweck ist nicht Kosmetik: Werden alle Druckermodelle installiert,
/// dauert der Start 13,5 s und die Auswahllisten enthalten 5762
/// Filamente. Mit einem einzelnen gewaehlten Drucker sind es 1,9 s und
/// 189 Filamente. Die Gruppierung nach Familien kommt aus dem
/// gemeinsamen Modul.
@Composable
fun SetupView(
    models: List<PsmCore.PrinterModel>,
    busy: Boolean,
    /** Bisher installierte Modelle, Format "vendor:model:variant". */
    preselected: Set<String>,
    onConfirm: (List<String>) -> Unit,
    onLanguageChange: (String) -> Unit,
    onClose: (() -> Unit)? = null,
) {
    val ps = LocalPsScale.current
    var selected by remember { mutableStateOf<Set<String>>(emptySet()) }
    var query by remember { mutableStateOf("") }
    var expandedVendors by remember { mutableStateOf<Set<String>>(emptySet()) }
    var expandedFamilies by remember { mutableStateOf<Set<String>>(emptySet()) }

    /// Ist es eng? Dann faellt Beiwerk weg, statt alles zu schrumpfen.
    val tight = ps.factor <= 0.8f

    val shown = models.filter { m ->
        // Hart aus, nicht nur standardmaessig versteckt: PSMobile
        // schneidet nur FFF, und das SLA-Buendel ist zudem kaputt.
        !m.isSla &&
            (query.trim().isEmpty() || m.name.contains(query, ignoreCase = true))
    }

    /// Zwei Ebenen: Hersteller, darunter Familien - beide Einteilungen
    /// aus dem gemeinsamen Modul.
    val hersteller: List<Hersteller> = run {
        val liste = shown
        PrinterGrouping.groupByVendor(liste.map { it.key }).map { v ->
            val modelle = v.indices.map { liste[it] }
            val familien = PrinterGrouping.group(modelle.map { it.family }).map { g ->
                Familie(
                    name = g.family,
                    isLegacy = g.isLegacy,
                    modelle = g.indices.map { modelle[it] },
                )
            }
            Hersteller(name = v.family, familien = familien)
        }
    }

    // Beim Erscheinen und noch einmal, wenn die gespeicherte Wahl
    // spaeter hereinkommt - siehe die gleichlautende Stelle in
    // SetupView.swift. Nur solange nichts gewaehlt ist: eine getroffene
    // Wahl darf eine nachziehende Voreinstellung nicht ueberschreiben.
    //
    // Hier ist es Vorsorge, drueben war es ein Reparaturversuch: dort
    // hatte die Ansicht die alte Druckerwahl uebernommen, bevor
    // -psm-reset-setup sie leerte. Seit dem 11.09.2026 wirkt der
    // Schalter drueben in PSMobileApp.init(), vor dem ersten Bild -
    // hier tut er es seit jeher in SlicerService.ensureCore.
    LaunchedEffect(preselected) { if (selected.isEmpty()) selected = preselected }

    SchwebenderDialog(kennung = "dialog.einrichtung", maximaleBreite = ps.pt(760)) {
        Column(
            Modifier
                .fillMaxHeight()
                .padding(ps.pt(if (tight) 12 else 24)),
            verticalArrangement = Arrangement.spacedBy(ps.pt(if (tight) 6 else 14)),
            horizontalAlignment = Alignment.Start,
        ) {
            Kopfzeile(tight, onClose)
            Suchfeld(query, onQuery = { query = it })
            Liste(
                modifier = Modifier.weight(1f),
                alle = hersteller,
                query = query,
                tight = tight,
                expandedVendors = expandedVendors,
                expandedFamilies = expandedFamilies,
                selected = selected,
                onVendor = { name ->
                    expandedVendors = if (name in expandedVendors) expandedVendors - name else expandedVendors + name
                },
                onFamilie = { name ->
                    expandedFamilies = if (name in expandedFamilies) expandedFamilies - name else expandedFamilies + name
                },
                onVariante = { key ->
                    selected = if (key in selected) selected - key else selected + key
                },
            )
            Abschluss(tight, busy, selected, onConfirm)
        }
    }
}

/// Eine Familie mit ihren Modellen.
private data class Familie(
    val name: String,
    val isLegacy: Boolean,
    val modelle: List<PsmCore.PrinterModel>,
)

/// Ein Hersteller mit seinen Familien.
private data class Hersteller(
    val name: String,
    val familien: List<Familie>,
) {
    val anzahl: Int get() = familien.sumOf { it.modelle.size }
}

// MARK: - Teile

@Composable
private fun Kopfzeile(tight: Boolean, onClose: (() -> Unit)?) {
    val ps = LocalPsScale.current
    Row(
        Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(ps.pt(8)),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (onClose != null) {
            Text(
                "‹  " + st("Back", "Zurück"),
                Modifier
                    .clickable(onClick = onClose)
                    .testTag("einrichtung.schliessen"),
                fontSize = ps.font(15), color = PrusaColors.orange,
            )
        }
        Column(verticalArrangement = Arrangement.spacedBy(ps.pt(2)), horizontalAlignment = Alignment.Start) {
            Text(
                st("Configuration Assistant", "Ersteinrichtung"),
                fontSize = ps.font(if (tight) 18 else 22), fontWeight = FontWeight.SemiBold,
                color = PrusaColors.textPrimary,
            )
            // Auf engen Schirmen faellt der Untertitel weg.
            if (!tight) {
                Text(
                    st("Select all printers you want to use.", "Wähle alle Drucker, die du nutzen willst."),
                    fontSize = ps.font(13), color = PrusaColors.textMuted,
                )
            }
        }
        Spacer(Modifier.weight(1f))
    }
}

@Composable
private fun Suchfeld(query: String, onQuery: (String) -> Unit) {
    val ps = LocalPsScale.current
    // Der Platzhalter wird selbst gezeichnet, in einer Farbe, die auf
    // dem dunklen Feld auch zu sehen ist.
    Box(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(ps.pt(6)))
            .background(PrusaColors.panelRaised),
        contentAlignment = Alignment.CenterStart,
    ) {
        if (query.isEmpty()) {
            Text(
                st("Search printer model", "Druckermodell suchen"),
                Modifier.padding(horizontal = ps.pt(12)),
                fontSize = ps.font(15), color = PrusaColors.textMuted,
            )
        }
        BasicTextField(
            value = query,
            onValueChange = onQuery,
            modifier = Modifier
                .fillMaxWidth()
                .padding(ps.pt(12))
                // Bis zum 11.09.2026 namenlos: die Tests griffen es als
                // "das erste Textfeld". Das haelt genau so lange, bis
                // ein zweites dazukommt.
                .testTag("einrichtung.suche"),
            textStyle = TextStyle(fontSize = ps.font(15), color = PrusaColors.textPrimary),
            cursorBrush = SolidColor(PrusaColors.orange),
            singleLine = true,
        )
    }
}

@Composable
private fun Liste(
    modifier: Modifier,
    alle: List<Hersteller>,
    query: String,
    tight: Boolean,
    expandedVendors: Set<String>,
    expandedFamilies: Set<String>,
    selected: Set<String>,
    onVendor: (String) -> Unit,
    onFamilie: (String) -> Unit,
    onVariante: (String) -> Unit,
) {
    // Die Liste traegt einen Namen, damit ein Test zu einem Hersteller
    // blaettern kann. Ohne das sieht er nur, was gerade gezeichnet ist -
    // und PrusaResearch steht alphabetisch weit unten.
    LazyColumn(modifier.fillMaxWidth().testTag("einrichtung.liste")) {
        alle.forEach { marke ->
            // Bei nur einem Hersteller waere die oberste Ebene eine
            // Zeile, die man immer erst aufklappen muss, ohne dass sie
            // etwas unterscheidet.
            val offen = alle.size == 1 ||
                expandedVendors.contains(marke.name) ||
                query.isNotEmpty()
            if (alle.size > 1) {
                item(key = "hersteller:${marke.name}") {
                    HerstellerKopf(marke, offen, onVendor)
                }
            }
            if (offen) {
                marke.familien.forEach { gruppe ->
                    item(key = "familie:${marke.name}/${gruppe.name}") {
                        FamilienKopf(gruppe, expandedFamilies.contains(gruppe.name), onFamilie)
                    }
                    if (expandedFamilies.contains(gruppe.name) || query.isNotEmpty()) {
                        items(gruppe.modelle, key = { "modell:${it.key}" }) { modell ->
                            ModellZeile(modell, tight, selected, onVariante)
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun HerstellerKopf(marke: Hersteller, offen: Boolean, onVendor: (String) -> Unit) {
    val ps = LocalPsScale.current
    Row(
        Modifier
            .fillMaxWidth()
            .clickable { onVendor(marke.name) }
            .testTag("hersteller.${marke.name}")
            .padding(top = ps.pt(8))
            .heightIn(min = ps.touch(52)),
        horizontalArrangement = Arrangement.spacedBy(ps.pt(8)),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(marke.name, fontSize = ps.font(17), fontWeight = FontWeight.Bold, color = PrusaColors.textPrimary)
        Text("${marke.anzahl}", fontSize = ps.font(13), color = PrusaColors.textMuted)
        Spacer(Modifier.weight(1f))
        SfSymbol(
            if (offen) "chevron.down" else "chevron.right",
            Modifier.size(ps.pt(14)),
            tint = PrusaColors.textMuted,
        )
    }
}

@Composable
private fun FamilienKopf(gruppe: Familie, offen: Boolean, onFamilie: (String) -> Unit) {
    val ps = LocalPsScale.current
    Row(
        Modifier
            .fillMaxWidth()
            .clickable { onFamilie(gruppe.name) }
            // Kennungen, damit der UI-Test die Zeilen findet.
            .testTag("familie.${gruppe.name}")
            .heightIn(min = ps.touch(52)),
        horizontalArrangement = Arrangement.spacedBy(ps.pt(8)),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            gruppe.name.uppercase(),
            Modifier.padding(start = ps.pt(12)),
            fontSize = ps.font(14), fontWeight = FontWeight.SemiBold,
            color = if (gruppe.isLegacy) PrusaColors.textMuted else PrusaColors.orange,
        )
        Text("${gruppe.modelle.size}", fontSize = ps.font(13), color = PrusaColors.textMuted)
        Spacer(Modifier.weight(1f))
        SfSymbol(
            if (offen) "chevron.down" else "chevron.right",
            Modifier.size(ps.pt(14)),
            tint = PrusaColors.textMuted,
        )
    }
}

@Composable
private fun ModellZeile(
    modell: PsmCore.PrinterModel,
    tight: Boolean,
    selected: Set<String>,
    onVariante: (String) -> Unit,
) {
    val ps = LocalPsScale.current
    // Jede Duesengroesse ist eine eigene Wahl - PrusaSlicer fuehrt sie
    // als getrennte Profile.
    Row(
        Modifier
            .fillMaxWidth()
            .padding(vertical = ps.pt(if (tight) 10 else 14)),
        horizontalArrangement = Arrangement.spacedBy(ps.pt(10)),
        verticalAlignment = Alignment.Top,
    ) {
        DruckerBild(modell, tight)

        Column(verticalArrangement = Arrangement.spacedBy(ps.pt(4)), horizontalAlignment = Alignment.Start) {
            Text(modell.name, fontSize = ps.font(if (tight) 16 else 17), color = PrusaColors.textPrimary)

            Row(horizontalArrangement = Arrangement.spacedBy(ps.pt(6))) {
                modell.variants.forEach { variante ->
                    val key = "${modell.key}:$variante"
                    val gewaehlt = selected.contains(key)
                    Box(
                        Modifier
                            .clip(RoundedCornerShape(ps.pt(4)))
                            .background(if (gewaehlt) PrusaColors.orange else PrusaColors.panelRaised)
                            .clickable { onVariante(key) }
                            .testTag("variante.$key")
                            .widthIn(min = ps.touch(64))
                            .heightIn(min = ps.touch(if (tight) 48 else 56))
                            .padding(horizontal = ps.pt(16)),
                        contentAlignment = Alignment.Center,
                    ) {
                        Text(
                            variante,
                            fontSize = ps.font(15),
                            color = if (gewaehlt) Color.White else PrusaColors.textMuted,
                        )
                    }
                }
            }
        }
    }
}

/// Vorschaubild je Druckermodell, wie in der Desktop-App - liegt als
/// loses PNG in den entpackten Ressourcen (profiles/<vendor>/
/// <model_id>_thumbnail.png). Fehlt die Datei, bleibt ein
/// Platzhalter-Symbol statt einer leeren Luecke.
@Composable
private fun DruckerBild(modell: PsmCore.PrinterModel, tight: Boolean) {
    val ps = LocalPsScale.current
    val context = LocalContext.current
    val seite = ps.pt(if (tight) 40 else 48)
    val bild by produceState<ImageBitmap?>(initialValue = null, key1 = modell.key) {
        value = withContext(Dispatchers.IO) {
            runCatching {
                thumbnailPfad(context, modell)?.let { BitmapFactory.decodeFile(it)?.asImageBitmap() }
            }.getOrNull()
        }
    }
    Box(
        Modifier
            .size(seite)
            .clip(RoundedCornerShape(ps.pt(4)))
            .background(PrusaColors.panelRaised),
        contentAlignment = Alignment.Center,
    ) {
        val b = bild
        if (b != null) {
            Image(b, contentDescription = null, modifier = Modifier.fillMaxSize(), contentScale = ContentScale.Fit)
        } else {
            SfSymbol("printer", Modifier.size(ps.pt(18)), tint = PrusaColors.textMuted)
        }
    }
}

/** Einmal nachgesehen, dann gemerkt - ein Aufruf je Zeile waere Laerm im Protokoll. */
@Volatile private var ressourcenWurzel: File? = null

private fun thumbnailPfad(context: Context, modell: PsmCore.PrinterModel): String? {
    // modell.key hat die Form "vendor:model_id" - der Dateiname der
    // Vorschau folgt bei den meisten Herstellern genau der model_id.
    // Voron ist die Ausnahme ("Voron_v2_250_thumbnail.png" statt
    // "V2_250_thumbnail.png"). Deshalb erst den genauen Namen
    // versuchen, sonst im Vendor-Ordner nach einer "*_thumbnail.png"
    // suchen, die die model_id enthaelt.
    val teile = modell.key.split(":", limit = 2)
    if (teile.size != 2) return null
    val resDir = ressourcenWurzel
        ?: runCatching { ResourceInstaller.ensureInstalled(context) }.getOrNull()?.also { ressourcenWurzel = it }
        ?: return null
    val vendorDir = File(File(resDir, "profiles"), teile[0])
    val genau = File(vendorDir, "${teile[1]}_thumbnail.png")
    if (genau.isFile) return genau.path
    val modelId = teile[1].lowercase()
    return vendorDir.listFiles()?.firstOrNull {
        val n = it.name.lowercase()
        n.endsWith("_thumbnail.png") && n.contains(modelId)
    }?.path
}

@Composable
private fun Abschluss(
    tight: Boolean,
    busy: Boolean,
    selected: Set<String>,
    onConfirm: (List<String>) -> Unit,
) {
    val ps = LocalPsScale.current
    Box(
        Modifier
            .fillMaxWidth()
            .height(ps.touch(if (tight) 46 else 58))
            .clip(RoundedCornerShape(ps.pt(6)))
            .background(if (selected.isEmpty()) PrusaColors.panelRaised else PrusaColors.orange)
            .clickable(enabled = !(selected.isEmpty() || busy)) { onConfirm(selected.toList()) }
            .testTag("fertig"),
        contentAlignment = Alignment.Center,
    ) {
        if (busy) {
            CircularProgressIndicator(Modifier.size(ps.pt(22)), color = Color.White, strokeWidth = 2.dp)
        } else {
            Text(
                st("Finish", "Fertig"),
                fontSize = ps.font(16), fontWeight = FontWeight.SemiBold,
                color = if (selected.isEmpty()) PrusaColors.textMuted else Color.White,
            )
        }
    }
}
